src/main/java/games/strategy/util/OpenJsonUtils.java

```diff
@@ -0,0 +1,119 @@
+   */
+  public static <T extends Enum<T>> T optEnum(
+      final JSONObject jsonObject,
+      final Class<T> type,
+      final String name,
+      final T defaultValue) {
+    checkNotNull(jsonObject);
+    checkNotNull(type);
+    checkNotNull(name);
+    checkNotNull(defaultValue);
+
+    final String valueName = jsonObject.optString(name);
+    return valueName.isEmpty() ? defaultValue : Enum.valueOf(type, valueName);
+  }
+
+  /**
+   * Streams the elements of the specified JSON array.
+   *
+   * @param jsonArray The JSON array.
+   *
+   * @return A stream containing the elements of the specified JSON array.
+   */
+  public static Stream<Object> stream(final JSONArray jsonArray) {
+    checkNotNull(jsonArray);
+
+    final int size = jsonArray.length();
+    final List<Object> list = new ArrayList<>(size);
+    for (int i = 0; i < size; ++i) {
+      list.add(jsonArray.get(i));
+    }
+    return list.stream();
+  }
+
+  /**
+   * Converts the specified JSON array into a list of equivalent Java types. Any {@link JSONArray} element will be
+   * converted into an equivalent {@code List<Object>}; any {@link JSONObject} element will be converted into an
+   * equivalent {@code Map<String, Object>}.
+   *
+   * @param jsonArray The JSON array.
+   *
+   * @return A list of the specified JSON array's elements converted to their equivalent Java types.
+   */
+  public static List<Object> toList(final JSONArray jsonArray) {
+    checkNotNull(jsonArray);
+
+    final int size = jsonArray.length();
+    final List<Object> list = new ArrayList<>(size);
+    for (int i = 0; i < size; ++i) {
+      list.add(unwrap(jsonArray.get(i)));
+    }
+    return list;
+  }
+
+  private static @Nullable Object unwrap(final @Nullable Object value) {
+    if (value == null || JSONObject.NULL.equals(value)) {
+      return null;
+    } else if (value instanceof JSONArray) {
+      return toList((JSONArray) value);
+    } else if (value instanceof JSONObject) {
+      return toMap((JSONObject) value);
+    }
+    return value;
+  }
+
+  /**
+   * Converts the properties of the specified JSON object into a map of equivalent Java types. Any {@link JSONArray}
+   * property value will be converted into an equivalent {@code List<Object>}; any {@link JSONObject} property value
+   * will be converted into an equivalent {@code Map<String, Object>}.
+   *
+   * @param jsonObject The JSON object.
+   *
+   * @return A map of the specified JSON object's properties converted to their equivalent Java types. The keys are the
+   *         property names; the values are the property values.
+   */
+  public static Map<String, Object> toMap(final JSONObject jsonObject) {
+    checkNotNull(jsonObject);
+
+    final Map<String, Object> map = new HashMap<>(jsonObject.length());
+    for (final String name : jsonObject.keySet()) {
+      map.put(name, unwrap(jsonObject.get(name)));
```

Comment:
Streaming works here as well, although not that pretty ^^
```java
return jsonObject.keySet().stream()
    .collect(Collectors.toMap(
        Function.indentity(),
        name -> unwrap(jsonObject.get(name))));
```

Reply:
Even better:
```java
final Function<String, Object> keyToValue = jsonObject::get;
return jsonObject.keySet().stream()
    .collect(Collectors.toMap(
        Function.indentity(),
        keyToValue.andThen(OpenJsonUtils::unwrap)));
// Alternatively
return jsonObject.keySet().stream()
    .collect(Collectors.toMap(
        Function.indentity(),
        ((Function<String, Object>)jsonObject::get)).andThen(OpenJsonUtils::unwrap)));
```

Reply:
Unfortunately, `Collectors#toMap()` calls `Map#merge()`, which doesn't seem to like `null` values (which are legal based on the json.org library's API):

```
java.lang.NullPointerException
	at java.util.HashMap.merge(HashMap.java:1224)
	at java.util.stream.Collectors.lambda$toMap$58(Collectors.java:1320)
	at java.util.stream.ReduceOps$3ReducingSink.accept(ReduceOps.java:169)
	at java.util.Iterator.forEachRemaining(Iterator.java:116)
	at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
	at games.strategy.util.OpenJsonUtils.toMap(OpenJsonUtils.java:116)
	at games.strategy.util.OpenJsonUtilsTest.toMap_ShouldConvertNullSentinelValueToNull(OpenJsonUtilsTest.java:174)
	<...snip...>
```

Reply:
Nevermind then